### PR TITLE
More array and hash methods

### PR DESCRIPTION
**Pull request by stephaniewillhide**
#46

### CODE REVIEW

File: app/models/author.rb [hunk 1]
```diff
@@ -35,6 +37,43 @@ def leap_year?
...
+      "Cancer ♋️" => [Date.parse("June 21 #{ date_of_birth.year }"), Date.parse("July 22 #{ date_of_birth.year }")],
+      "Leo ♌️" => [Date.parse("July 23 #{ date_of_birth.year }"), Date.parse("August 22 #{ date_of_birth.year }")],
+      "Virgo ♍️" => [Date.parse("August 23 #{ date_of_birth.year }"), Date.parse("September 22 #{ date_of_birth.year }")],
+      "Libra ♎️" => [Date.parse("September 23 #{ date_of_birth.year }"), Date.parse("October 22 #{ date_of_birth.year }")],
+      "Scorpio ♏️" => [Date.parse("October 23 #{ date_of_birth.year }"), Date.parse("November 21 #{ date_of_birth.year }")],
+      "Sagittarius ♐️" => [Date.parse("November 22 #{ date_of_birth.year }"), Date.parse("December 21 #{ date_of_birth.year }")],
+      # "Capricorn ♑️" => [Date.parse("December 22 #{ date_of_birth.year }"), Date.parse("January 19 #{ date_of_birth.year }")],
+      "Aquarius ♒️" => [Date.parse("January 20 #{ date_of_birth.year }"), Date.parse("February 18 #{ date_of_birth.year }")],
+      "Pisces ♓️" => [Date.parse("February 19 #{ date_of_birth.year }"), Date.parse("March 20 #{ date_of_birth.year }")],
+    }
```
Comment (jordanmaguire) [inline app/models/author.rb:54]: There's a lot of repetition in this block, with plenty of what the reader is ingesting being boilerplate code (EG: Date.parse, date_of_birth.year).

We should aim to reduce the amount of boilerplate code we need to read, so that we can focus on what matters.

In this case, we really want to focus primarily on the day and months.

We could use a lambda to achieve this:

```ruby
date_for = -> (month_and_day) { Date.parse("#{ month_and_day } #{ date_of_birth.year }") }

zodiacs = {
  "Aries ♈️" => [date_for.call("March 21"), date_for.call("April 19")],
  "Taurus ♉️" => [date_for.call("April 20"), date_for.call("May 20")],
  "Gemini ♊️" => [date_for.call("May 21"), date_for.call("June 20")],
  "Cancer ♋️" => [date_for.call("June 21"), date_for.call("July 22")],
  "Leo ♌️" => [date_for.call("July 23"), date_for.call("August 22")],
  "Virgo ♍️" => [date_for.call("August 23"), date_for.call("September 22")],
  "Libra ♎️" => [date_for.call("September 23"), date_for.call("October 22")],
  "Scorpio ♏️" => [date_for.call("October 23"), date_for.call("November 21")],
  "Sagittarius ♐️" => [date_for.call("November 22"), date_for.call("December 21")],
  # "Capricorn ♑️" => [date_for.call("December 22"), date_for.call("January 19")],
  "Aquarius ♒️" => [date_for.call("January 20"), date_for.call("February 18")],
  "Pisces ♓️" => [date_for.call("February 19"), date_for.call("March 20")],
}
```

Or, we could take another path entirely, and move the date parsing somewhere else:

```ruby
zodiacs = {
  "Aries ♈️" => ["March 21", "April 19"],
  "Taurus ♉️" => ["April 20", "May 20"],
  "Gemini ♊️" => ["May 21", "June 20"],
  "Cancer ♋️" => ["June 21", "July 22"],
  "Leo ♌️" => ["July 23", "August 22"],
  "Virgo ♍️" => ["August 23", "September 22"],
  "Libra ♎️" => ["September 23", "October 22"],
  "Scorpio ♏️" => ["October 23", "November 21"],
  "Sagittarius ♐️" => ["November 22", "December 21"],
  # "Capricorn ♑️" => ["December 22", "January 19"],
  "Aquarius ♒️" => ["January 20", "February 18"],
  "Pisces ♓️" => ["February 19", "March 20"],
}

zodiac_array = zodiacs.find do |zodiac, (begin_month_and_day, end_month_and_day)|
  begin_date = Date.parse("#{ begin_month_and_day } #{ date_of_birth.year }")
  end_date = Date.parse("#{ end_month_and_day } #{ date_of_birth.year }")

  date_of_birth.between?(begin_date, end_date)
end
```

File: spec/models/author_spec.rb [hunk 2]
```diff
@@ -129,4 +129,37 @@
...
+        "Gemini ♊️" => [Date.parse("May 21 2002"), Date.parse("June 20 2002")],
+        "Cancer ♋️" => [Date.parse("June 21 2002"), Date.parse("July 22 2002")],
+        "Leo ♌️" => [Date.parse("July 23 2002"), Date.parse("August 22 2002")],
+        "Virgo ♍️" => [Date.parse("August 23 2002"), Date.parse("September 22 2002")],
+        "Libra ♎️" => [Date.parse("September 23 2002"), Date.parse("October 22 2002")],
+        "Scorpio ♏️" => [Date.parse("October 23 2002"), Date.parse("November 21 2002")],
+        "Sagittarius ♐️" => [Date.parse("November 22 2002"), Date.parse("December 21 2002")],
+        # "Capricorn ♑️" => [Date.parse("December 22 2002"), Date.parse("January 19 2002")],
+        "Aquarius ♒️" => [Date.parse("January 20 2002"), Date.parse("February 18 2002")],
+        "Pisces ♓️" => [Date.parse("February 19 2002"), Date.parse("March 20 2002")],
```
Comment (jordanmaguire) [inline spec/models/author_spec.rb:148]: You could save some boilerplate code here by parsing the date inside the block, then you'd just have the date string here and not need to write Date.parse 20 times

File: app/models/author.rb (hunk 1 above)
Comment (stephaniewillhide) [reply app/models/author.rb:54]: I prefer the second option! It seems more intuitive to me. I'm going to implement this.

File: app/models/zodiac.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  def include?(date)
+    # Capricorn spans the beginning and end of a year. For all other dates we can compare the start
+    # and end dates.
+    begin_date = Date.parse("#{ begin_month_and_day } #{ date.year }")
+    end_date = Date.parse("#{ end_month_and_day } #{ date.year }")
+    if name == "Capricorn ♑️"
+      date >= begin_date || date <= end_date
+    else
+      date.between?(begin_date, end_date)
+    end
```
Comment (jordanmaguire) [inline app/models/zodiac.rb:40]: What reading your tests made me think about was that the identity of the Zodiac (the name combined with the start and end date) wasn't really important.

So the check on the name of the zodiac is flimsy, because if another zodiac was added that spanned the year, our implementation would break as Capricornicus would no longer be the relevant zodiac.

Instead, a more accurate way to write this that is future-proof is to check whether the start and end dates are after eachother.

EG:

```ruby
if end_date < begin_date
   date >= begin_date || date <= end_date
else
  date.between?(begin_date, end_date)
end
```

Now your tests can be more effective, because we have more clear edge cases.

File: spec/models/author_spec.rb [hunk 4]
```diff
@@ -129,4 +129,53 @@
...
+      }
+
+      # Because Capricorn spans 2 years, a comparison is more complex than the other zodiacs
+      # To ensure it is properly tested, we are testing explicitly below
+
+      ["2002-01-01", "2002-01-19", "2002-12-22", "2002-12-31"].each { |date|
+        capricorn_author = create(:author, date_of_birth: date)
+        expect(capricorn_author.zodiac).to eq("Capricorn ♑️")
+      }
+    end
```
Comment (jordanmaguire) [inline spec/models/author_spec.rb:163]: At this point in the refactor, we should attempt to use stubs to make the tests for this zodiac method shorter.

We don't want to duplicate the sample data in every test that relies on the Zodiac class.

File: spec/models/author_spec.rb [hunk 5]
```diff
@@ -129,4 +129,53 @@
...
+  end
+
+  describe ".age_at_first_release" do
+    it "Calculates age at first release" do
+      author = create(:author, date_of_birth: "Sun, 15 Dec 1980")
```
Comment (jordanmaguire) [inline spec/models/author_spec.rb:168]: I'd try and keep the data format you use consistent, that way it's easy to know which format you should use in future, and it's easier for people reading the code.

**Review by jordanmaguire — COMMENTED** [review]
I've put some preliminary comments in here, if you can tackle those I'll take another look

File: spec/models/author_spec.rb (hunk 4 above)
Comment (stephaniewillhide) [reply spec/models/author_spec.rb:163]: I've pushed up what I have here, but I can tell that it's not right because my tests are failing. The specific error I'm getting here is `undefined method `include?' for :zodiac:Symbol`. `include?` is defined, so I'm assuming that the way I'm calling the Zodiac class in the stubs is likely the problem.

File: app/models/author.rb [hunk 6]
```diff
@@ -35,6 +37,20 @@ def leap_year?
...
+  end
+
+  def age_at_first_release
+    date_of_first_book = books.map do |book|
+      book.published_at
+    end.min
+
+    diff = ((date_of_first_book - date_of_birth).to_i)/365.25
+
+    "About #{ diff.floor } years and #{ (diff.modulo(1)*365.25).round(0) } days"
```
Comment (jordanmaguire) [inline app/models/author.rb:51]: given we use `365.25` twice here, I'd be inclined to pop it in a self-documenting local variable. EG:

```ruby
average_number_of_days_in_a_year = 365.25
diff = (date_of_first_book - date_of_birth).to_i / average_number_of_days_in_a_year
```

I'd also be inclined to extract some of the interpolated values in the string, again for the purpose of self documenting and because I personally find calculations in interpolations difficult to read

```ruby
years = diff.floor
days = (diff.modulo(1) * average_number_of_days_in_a_year).round(0)

"About #{ years } years and #{ days } days"
```

File: app/models/author.rb [hunk 7]
```diff
@@ -35,6 +37,20 @@ def leap_year?
...
   end
 
+  def zodiac
+    Zodiac.for_date(date_of_birth)
+  end
+
+  def age_at_first_release
+    date_of_first_book = books.map do |book|
+      book.published_at
+    end.min
```
Comment (jordanmaguire) [inline app/models/author.rb:47]: ruby has a shorthand for generating blocks that can make things more readable. In this case:

```ruby
date_of_first_book = books.map(&:published_at).min
```

**Review by jordanmaguire — APPROVED** [review]
